### ISSUE TITLE
imap4-utf-7 codec improperly registered

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3663]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3663|
|Type|defect|
|Created|2009-02-21 16:23:57Z|
|Branch|https://github.com/twisted/twisted/tree/imap4-utf7-registration-3663|

In Python 2.5, it's not possible to look up the imap4-utf-7 codec registered by twisted/mail/imap4.py:

```
exarkun@boson:~$ python
Python 2.5.2 (r252:60911, Jul 31 2008, 17:28:52)
[GCC 4.2.3 (Ubuntu 4.2.3-2ubuntu7)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from twisted.mail import imap4
>>> import codecs
>>> codecs.lookup('imap4-utf-7')
(<function encoder at 0x8395df4>, <function decoder at 0x8395e2c>, <class twisted.mail.imap4.StreamReader at 0x837bdac>, <class twisted.mail.imap4.StreamWriter at 0x837bddc>)
>>> codecs.getreader('imap4-utf-7')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.5/codecs.py", line 920, in getreader
    return lookup(encoding).streamreader
AttributeError: 'tuple' object has no attribute 'streamreader'
>>> 
```

Though this did use to work:

```
exarkun@boson:~$ ~/Projects/python/branches/release23-maint/python 
Python 2.3.7 (#2, Mar 25 2008, 19:31:49) 
[GCC 4.1.3 20070929 (prerelease) (Ubuntu 4.1.2-16ubuntu2)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from twisted.mail import imap4
>>> import codecs
>>> codecs.getreader('imap4-utf-7')
<class twisted.mail.imap4.StreamReader at 0xb7a0d5fc>
>>> 
```

Also, `StreamWriter` might be misdefined.  It has a `decode` method, but perhaps it should have an `encode` method.

<details><summary>Searchable metadata</summary>

```
trac-id__3663 3663
type__defect defect
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__branches_imap4_utf7_registration_3663 branches/imap4-utf7-registration-3663
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__easy easy
time__1235233437000000 1235233437000000
changetime__1237565316000000 1237565316000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [26379]) Branching to 'imap4-utf7-registration-3663'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [26380]) Fix the definition of StreamWriter and return a CodecInfo from the lookup function

refs [#3663](https://github.com/twisted/twisted/issues/3663)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/imap4-utf7-registration-3663)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @exarkun|
|-|-|

Looks good.  Merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [26479]) Merge imap4-utf7-registration-3663

Author: exarkun
Reviewer: glyph
Fixes: [#3663](https://github.com/twisted/twisted/issues/3663)

Update the IMAP4-UTF7 codec registration to use the preferred Python 2.5 behavior
of returning a CodecInfo instance instead of a four tuple.  This causes the
codecs.getreader and codecs.getwriter functions to work for this encoding (again).
Also fix the implementation of the stream writer to define the correct method,
`encode`, instead of the incorrect one, `decode`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
